Additional crystals added to the EcalBadCal MET filter

lathomas opened the pull request:
New hot EE crystals have been recently observed in 2017 and 2018. 
After discussion between ECAL and JME it was decided to add them to the list of crystal considered by the ecal bad calibration MET filter. 

More information here: 
https://indico.cern.ch/event/775564/contributions/3239371/attachments/1765837/2866952/ecalbadcalfilterupdate.pdf
https://twiki.cern.ch/twiki/bin/viewauth/CMS/MissingETOptionalFiltersRun2#How_to_run_ecal_BadCalibReducedM 

@ahinzmann @peruzzim @eioannou @zdemirag

lathomas committed:
Adding additional crystals to the list of hot crystals checked by the filter

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25436/7501

cmsbuild commented:
A new Pull Request was created by  @lathomas  for master.

It involves the following packages:

RecoMET/METFilters


@perrotta, @monttj, @cmsbuild, @fgolf, @slava77, @peruzzim can you please review it and eventually sign? Thanks.
@TaiSakuma, @gouskos, @ahinzmann, @mmarionncern, @jdamgov, @jdolen, @nhanvtran, @gkasieczka, @clelange, @schoef, @mariadalfonso, @seemasharmafnal this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

peruzzim commented:
Please ask for advice and agree on how to extend the crystal list also in view of backport to 10_2_X, where you will not be allowed to just extend it like that because it would change the content of miniAOD.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

lathomas committed:
Duplicating the filter to allow it to rerun on MINIAOD and for 102X/94X backport.

cmsbuild commented:
The code-checks are being triggered in jenkins.

lathomas commented:
@peruzzim I updated the config file and added a new instance of the filter to be run on MINIAOD. 
From what I understand this can be safely backported to 102X and 94X. Can you confirm this is fine? 
If yes I'll open the PR.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25436/7505

cmsbuild commented:
Pull request #25436 was updated. @perrotta, @monttj, @cmsbuild, @fgolf, @slava77, @peruzzim can you please check and sign again.

peruzzim commented on RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py:
```diff
@@ -29,4 +29,39 @@
+                                  872437185,872422564,872421566,872421695,
+                                  872421955,872421567,872437184,872421951,
+                                  872421694,872437056,872437057,872437313]),
+  taggingMode = cms.bool(True),
```
I would suggest to set this to False by default, as it is for the filter above

peruzzim commented on RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py:
```diff
@@ -29,4 +29,39 @@
+                                                       872421694,872437056,872437057,872437313])
+
+
+ecalBadCalibReducedMINIAODFilter = cms.EDFilter(
```
I would suggest giving this a more peculiar name, as there is no a-priori reason to define bad crystals in a different way when running on AOD or MINIAOD.
I would name it ecalBadCalibReducedMINIAODPre104XFilter to make it clear it has to be run and considered only when running on MINIAOD produced in older releases.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
@peruzzim @fgolf 
this is not for nanoAOD production just yet, right?
Please confirm.
Thank you.

peruzzim commented:
@slava77 we were just going to comment, yes it would be very nice to have it.
Sorry again for the late notice, it is something that has emerged very recently.

slava77 commented:
> @slava77 we were just going to comment, yes it would be very nice to have it.

please clarify the meaning of this for this PR.
- the ecalBadCalibFilter runs only in miniAOD ==> will not be used in the upcoming 102X nanoAOD
- the new ecalBadCalibReducedMINIAODFilter is not added to any of the tasks/sequences in nanoAOD to create a usable variable.

So, presently this PR is not a part of nanoAOD processing

lathomas committed:
Updating filter name

peruzzim commented:
- ecalBadCalibFilter will not be used in 102X nanoAOD, but will have to be used from now on in productions to happen in 104X and later, with the updated list of bad xtals. So we are updating that filter to reject more xtals from now on.
- we want to add a filter cutting on the additional xtals in 102X nanoAOD, but we cannot just modify ecalBadCalibFilter because it would change the content of miniAOD in 102X (production). So we add an additional filter that will not be in the 102X miniAOD sequence, and will be called only from nanoAOD (PR will come as soon as we define its name).
- we have to support running on miniAOD produced in 103X and earlier releases, when running from a 104X area. So, in those cases, we want to run the filter on the additional xtals, on the reduced rechit collection. This could in principle be avoided by cloning ecalBadCalibFilter and changing the source collection, would you prefer that rather than having this filter defined in 104X+?

Only the update of ecalBadCalibReducedMINIAODFilter will be backported to 102X.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25436/7506

cmsbuild commented:
Pull request #25436 was updated. @perrotta, @monttj, @cmsbuild, @fgolf, @slava77, @peruzzim can you please check and sign again.

slava77 commented:
> that will not be in the 102X miniAOD sequence, and will be called only from nanoAOD (PR will come as soon as we define its name).

I prefer this to happen in one PR

cmsbuild commented:
+1
Tested at: 20e12a76d65e2a1c1400b00282f9bb1dfbc61ec3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25436/32040/summary.html

cmsbuild commented:
Comparison job queued.

peruzzim commented:
@slava77 sorry can you please detail how you would like us to do in order to support the use cases listed above?
What we would do in nanoAOD is pretty clear, we want to run the updated filter on the miniAOD xtal collection and save a bool with its result.
Can you please suggest your preferred implementation of this? Should we define a catch-all modifier like run2_miniaod_postAutumn18 and start to put all these things in there? I don't think we can define a modifier for every thing we're going to change in 102X in the next months.

slava77 commented on RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py:
```diff
@@ -29,4 +29,39 @@
+                                                       872421694,872437056,872437057,872437313])
+
+
+ecalBadCalibReducedMINIAODPre104XFilter = cms.EDFilter(
```
The settings should not be copy-pasted from what is in ecalBadCalibFilter.
Presently, ```ecalBadCalibReducedMINIAODPre104XFilter = ecalBadCalibFilter.clone(EcalRecHitSource = 'reducedEgamma:reducedEERecHits')``` should give identical result.

I don't understand the need to have "Pre104X" in the name. IIUC, the simplest is to just apply it to all data anyways on top of the default ecalBadCalibFilter. Either it's an analysis running on miniAOD made with the most up-to-date list of xtals [in this case this filter is redundant or just a subset of the full one ecalBadCalibFilter] or your miniAOD doesn't cover all xtals in which this ecalBadCalibReducedMINIAODFilter triggers (partially/the best way possible)  on the bad events.

Also, nominally, we do not allow multiple modules defined in cfi files.
So, this definition should be moved to a separate _cff file.

peruzzim commented:
Ok, agreed for 104X, thanks.
How should we do to backport this to 102X, preserving the default miniAOD behaviour there but allowing people (including nanoAOD producer, but also in general all analysers) to call it with the updated xtal list when running on miniAOD?
Thanks a lot for your help!

slava77 commented:
> @slava77 sorry can you please detail how you would like us to do in order to support the use cases listed above?
> What we would do in nanoAOD is pretty clear, we want to run the updated filter on the miniAOD xtal collection and save a bool with its result.
> Can you please suggest your preferred implementation of this? Should we define a catch-all modifier like run2_miniaod_postAutumn18 and start to put all these things in there? I don't think we can define a modifier for every thing we're going to change in 102X in the next months.

As we discussed in the RECO/AT meeting last week, there should be a "future miniAOD" modifier for all things proposed for a backport and modifying the default/available miniAOD results.
I was thinking ```run2_miniAOD_devel``` could be a good name.
This modifier will be empty in the master and will capture changes in not-yet-defined miniAOD in the production release.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25436/32040/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3131939
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3131734
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

slava77 commented:
On 12/6/18 11:21 AM, peruzzim wrote:
> Ok, agreed for 104X, thanks.
> How should we do to backport this to 102X, preserving the default
> miniAOD behaviour there but allowing people (including nanoAOD producer,
> but also in general all analysers) to call it with the updated xtal list
> when running on miniAOD?
> Thanks a lot for your help!
> 
> —

looking at what was done before,
I think that there should be something like run2_nanoAOD_102Xv1 modifier
for 2018 inputs
in addition to what you have for 94X miniAOD (the 2016 "v3" and 2017 "v2").
This shall formalize support of the 102X miniAOD format in the master branch
and can be used explicitly in the 102X backports as well.
Simply having everything in just Run2_2018 era will lead to problems for
simultaneous support in 102X and 104X/master because the miniAOD in the
master will be evolving.

peruzzim commented:
Thanks, we're doing as you suggest #25443 #25444.
An update of this PR, #25372 and #25368 will follow once the modifiers are merged.

lathomas committed:
Removing pre104 filter

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25436/32065/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3136422
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3136216
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

peruzzim commented:
@lathomas this PR needs to be also backported to 102X, modifying (only in the backport, the master looks good as it is) the list of xtals only when one of these modifiers is active:
- run2_miniAOD_devel
- run2_nanoAOD_102Xv1

the backport will depend on #25444

slava77 commented on RecoMET/METFilters/python/ecalBadCalibFilter_cfi.py:
```diff
@@ -29,4 +29,9 @@
                                                        872436657,872420397,872439732,872439339,
-                                                       872439603,872422436])
+                                                       872439603,872422436,872439861,872437051,
+                                                       872437052,872420649,872422436,872421950,
```
872422436 is showing up twice.
Please check that this is just a redundancy and not a typo missing some other detid.

lathomas replied:
@slava77  I just checked and I confirm that this is a redundancy, not a typo. This was the first "new" crystal spotted in 2018. We then converged on a more exhaustive list and for some reasons I copied this crystal twice. Sorry for that.

lathomas committed:
Removing duplicate crystal

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25436/7544

cmsbuild commented:
Pull request #25436 was updated. @perrotta, @monttj, @cmsbuild, @fgolf, @slava77, @peruzzim can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 550b24feb5abdc6a28100f4636e5da51d3cda7a3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25436/32095/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25436/32095/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3136422
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3136217
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/25436	550b24f
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences [some may be expected in the Flag_ecalBadCalibFilter decision, but the number of tested events is not that large to certainly expect a visible change]
